downstairs/src/repair.rs

```diff
@@ -0,0 +1,386 @@
+
+    if m.file_type().is_symlink() {
+        Err(HttpError::for_bad_request(None, "EMLINK".to_string()))
+    } else if path.is_dir() {
+        Err(HttpError::for_bad_request(None, "EBADF".to_string()))
+    } else {
+        let file = tokio::fs::File::open(&path).await.map_err(|_| {
+            HttpError::for_bad_request(None, "EBADF".to_string())
+        })?;
+
+        let file_stream = hyper_staticfile::FileBytesStream::new(file);
+        let content_type = "application/octet-stream".to_string();
+
+        Ok(Response::builder()
+            .status(StatusCode::OK)
+            .header(http::header::CONTENT_TYPE, content_type)
+            .body(file_stream.into_body())?)
+    }
+}
+
+#[derive(Deserialize, Serialize, JsonSchema)]
+struct ExtentFiles {
+    files: Vec<String>,
+}
+
+/**
+ * For a given extent, return a vec of strings representing the names of
+ * the files that exist for that extent.
+ */
+#[endpoint {
+    method = GET,
+    path = "/extent/{eid}/files",
+    unpublished = false,
+}]
+async fn get_files_for_extent(
+    rqctx: Arc<RequestContext<FileServerContext>>,
+    path: Path<Eid>,
+) -> Result<HttpResponseOk<ExtentFiles>, HttpError> {
+    let eid = path.into_inner().eid;
+    let extent_dir = extent_dir(rqctx.context().region_dir.clone(), eid);
+
+    // Some sanity checking on the extent path
+    let m = extent_dir
+        .symlink_metadata()
+        .map_err(|_| HttpError::for_bad_request(None, "ENOENT".to_string()))?;
+
+    if m.file_type().is_symlink() {
+        Err(HttpError::for_bad_request(None, "EMLINK".to_string()))
+    } else if !extent_dir.is_dir() {
+        Err(HttpError::for_bad_request(None, "EBADF".to_string()))
+    } else {
+        let files = extent_file_list(extent_dir, eid).await?;
+        Ok(HttpResponseOk(files))
+    }
+}
+
+async fn extent_file_list(
+    extent_dir: PathBuf,
+    eid: u32,
+) -> Result<ExtentFiles, HttpError> {
+    let mut full_name = extent_dir;
+
+    let extent_name = extent_file_name(eid, None);
+    full_name.push(extent_name.clone());
+    let mut files = Vec::new();
+    // The data file should always exist
+    if !full_name.exists() {
+        return Err(HttpError::for_bad_request(None, "EBADF".to_string()));
+    }
+    files.push(extent_name);
+    // The db file should always exist.
+    full_name.set_extension("db");
+    if !full_name.exists() {
+        return Err(HttpError::for_bad_request(None, "EBADF".to_string()));
+    }
+    files.push(extent_file_name(eid, Some("db")));
+
+    // The db-shm file may exist.
+    full_name.set_extension("db-shm");
+    if full_name.exists() {
+        println!("Exists: {:?}", full_name.file_name().unwrap());
+        files.push(extent_file_name(eid, Some("db-shm")));
+    }
+    // The db-wal file may exist.
+    full_name.set_extension("db-wal");
+    if full_name.exists() {
+        files.push(extent_file_name(eid, Some("db-wal")));
+    }
```

Comment:
```suggestion
    let files = &[
        (extent_file_name(eid, None), true),
        (extent_file_name(eid, Some("db")), true),
        (extent_file_name(eid, Some("db-shm"), false),
        (extent_file_name(eid, Some("db-wal"), false),
    ].into_iter()
    .filter_map(|(file, required)| {
        if file.exists() {
            Some(Ok(file))
        } else if required {
            Some(Err(HttpError::for_bad_request(None, "EBADF".to_string())))
        } else {
            None
        }
    })
    .collect::<Result<_>>()?;
```

I don't think this is perfect, but a sketch for your consideration. It may be more confusing than simplifying, but I wanted to try it out.